include/json/value.h

```diff
@@ -171,6 +171,9 @@ class JSON_API Value {
     CZString(ArrayIndex index);
     CZString(const char* cstr, DuplicationPolicy allocate);
     CZString(const CZString& other);
+#ifdef JSON_HAS_RVALUE_REFERENCES
+	CZString(CZString&& other);
+#endif
     ~CZString();
     CZString& operator=(CZString other);
```

Comment:
Move is the more fundamental operation. I think swap should really be built in terms of it rather than the other way round.

```c++
class Value {
    ...
    Value(const Value&);
    Value(Value&&) noexcept;
    Value& operator=(const Value&);
    Value& operator=(Value&&) noexcept;
    void swap(Value& v) noexcept { std::swap(*this, v); }
    friend void swap(Value& a, Value& b) noexcept { a.swap(b); }
};
```

The copy-and-swap trick is a pre-11 hack. It does extra work to fix up the moved-from source object only to delete it at the end of function. There are considerations of vector- or string capacity as well. A moved-from object can be better equipped for reuse if it is moved directly into its destination rather than moving into a temporary first.

I prefer separating the const T& and the T&& cases into separate functions in case they can be accomplished more efficiently with different strategies. Because of binary compatibility, we can't split them up later if we decide we want to. So really we need two functions.